tagstudio/src/core/library.py

```diff
@@ -2324,3 +2172,287 @@ def sort_fields(self, entry_id: int, order: list[int]) -> None:
+                        if entry_tuple is None:
+                            is_selected = False
+                            break
+                        else:
+                            is_selected = True
+                    # if entry does not have info on requested key
+                    elif select_empty and (entry_value is None or entry_value == ""):
+                        is_selected = True
+                    # check if entry's value match the query
+                    else:
+                        print(value, entry_value)
+                        if entry_value is not None and value in entry_value:
+                            print("true")
+                            is_selected = True
+                        else:
+                            is_selected = False
+                            break
+
+                if is_selected and entry_tuple is None:
+                    filtered_entries.append((ItemType.ENTRY, entry.id))
+                elif is_selected and entry_tuple is not None:
+                    filtered_entries.append(entry_tuple)
+
+            pre_results.append(filtered_entries)
+        # Entries should match all parts between '|'
+        result_set: set = set()
+        if search_mode == SearchMode.AND:
+            if len(pre_results) <= 1:
+                result_set = set(pre_results[0])
+                return list(pre_results[0])
+            result_set = set(pre_results[0])
+            # entries should be found in every part of query
+            for result in range(1, len(pre_results)):
+                set_copy: list = list(result_set.copy())
+                for recorded_entry in set_copy:
+                    entry_detected: bool = False
+                    for new_entry in pre_results[result]:
+                        if recorded_entry == new_entry:
+                            entry_detected = True
+                            break
+                    if not entry_detected:
+                        result_set.remove(recorded_entry)
+            return list(result_set)
+        # Entries should match any of parts between '|'
+        elif search_mode == SearchMode.OR:
+            for result in pre_results:  # type: ignore [assignment]
+                for entry in result:  # type: ignore [attr-defined]
+                    result_set.add(entry)
+            return list(result_set)
+
+    def handle_tag_id(self, query: str, entry_tags: list[int]) -> bool:
+        id_query: int = 0
+        try:
+            id_query = int(query)
+        except Exception:
+            return False
+        if id_query in entry_tags:
+            return True
+        return False
+
+    def handle_unbound(
+        self,
+        entry: Entry,
+        all_tag_terms: list[str],
+        entry_tags: list[int],
+        entry_authors: list[list[str]],
+        unbound: list[str],
+    ) -> tuple[ItemType, int] | None:
+        collations_added: list = []
+        orig_query = "".join(unbound)
```

Comment:
why is this converted to string? the conditions below would work even with list no? Also if you will add here this assert, you will see that some tests are passing here `unbound` as string instead of list. Which doesnt look correct according to the type annotation:

```suggestion
        assert isinstance(unbound, list), f'unbound is not list: {type(unbound)}'
        orig_query = "".join(unbound)
```

Reply:
They actually won't
``` python
>>> l: list = ['no', 'empty', 'artist']
>>> 'no artist' in l
False
>>> 'no empty' in l
False
```
The problem is that some special flags in the original version contained spaces, and we should be combing separate 'no' and 'artist' into 'no artist' this way. It could be done, but converting it to a monolithic string is done much easily and more readable in my opinion

Reply:
I see. And I dont know. I'm here just to review code for fun, but I dont make the product decision. But this, together with the difficult way how the negative queries like 'no artist' are parsed, makes me really think that the syntax -artist would be more straightforward and easier to work with.